lib/mesh.h

```diff
@@ -0,0 +1,337 @@
+        CornerVerticesProperty::createIfNotExists(mesh, "corner_vertices");
+
+    std::array<RadiosityMesh::VertexHandle, 3> vhandles;
+    for (const auto& tri : triangles) {
+        const Vec& a = tri.vertices[0];
+        const Vec& b = tri.vertices[1];
+        const Vec& c = tri.vertices[2];
+        const Vec& na = tri.normals[0];
+        const Vec& nb = tri.normals[1];
+        const Vec& nc = tri.normals[2];
+
+        auto ita = lookup.emplace(std::make_pair(a, na),
+                                  RadiosityMesh::VertexHandle());
+        if (ita.second) {
+            ita.first->second = mesh.add_vertex({a.x, a.y, a.z});
+        }
+        auto va = ita.first->second;
+
+        auto itb = lookup.emplace(std::make_pair(b, nb),
+                                  RadiosityMesh::VertexHandle());
+        if (itb.second) {
+            itb.first->second = mesh.add_vertex({b.x, b.y, b.z});
+        }
+        auto vb = itb.first->second;
+
+        auto itc = lookup.emplace(std::make_pair(c, nc),
+                                  RadiosityMesh::VertexHandle());
+        if (itc.second) {
+            itc.first->second = mesh.add_vertex({c.x, c.y, c.z});
+        }
+        auto vc = itc.first->second;
+
+        auto face = mesh.add_face(va, vb, vc);
+        corners_prop[face] = {va, vb, vc};
+    }
+
+    return mesh;
+}
+
+auto subdivide4(RadiosityMesh& mesh, RadiosityMesh::FaceHandle face) {
+    CornerVerticesProperty corner_vertices_prop(mesh, "corner_vertices", true);
+
+    // build index of corners and rearrange their coordinates accordingly
+    const auto corners = corner_vertices_prop[face];
+    std::array<RadiosityMesh::VertexHandle, 3>
+        corner_vertices;                // corner vertices
+    std::array<size_t, 3> corner_index; // corner indices
+    size_t index = 0;
+    size_t current = 0;
+    for (auto v : mesh.fv_range(face)) {
+        if (v == corners[0] || v == corners[1] || v == corners[2]) {
+            corner_index[current] = index;
+            corner_vertices[current] = v;
+            current++;
+        }
+        index++;
+    }
+    assert(current == 3);
+    size_t N = index;
+
+    size_t i0 = corner_index[0];
+    size_t i1 = corner_index[1];
+    size_t i2 = corner_index[2];
+
+    // prepare for addition of middle points
+    const auto& a = mesh.point(corner_vertices[0]);
+    const auto& b = mesh.point(corner_vertices[1]);
+    const auto& c = mesh.point(corner_vertices[2]);
+
+    const auto mab = (a + b) / 2;
+    const auto mbc = (b + c) / 2;
+    const auto mac = (a + c) / 2;
+
+    // add pt between i0 and i1
+    size_t k0 = detail::find_vertex(mesh, face, mab);
+    if (k0 == N) {
+        // no vertex found, i.e. neighbor face was not divided yet
+        k0 = i0 + 1;
+        auto he = detail::get_halfedge_handle(mesh, face, k0);
+        auto edge = mesh.edge_handle(he);
+        mesh.split(edge, mab);
+        N += 1;
+        i1 += 1;
+        i2 += 1;
```

Comment:
Ich will das nach dem Mergen refactorn. Man will nicht 3 mal nach einem Vertex suchen. Man macht es nämlich jedes mal von Anfang an. Das sollte in einer Iteration passieren.